### PR TITLE
fix: move main tab activation to puppeteer plugin

### LINKED ISSUE #28392: Cypress v13.6.0 no longer running commands (hangs) when we focus out of the window
### Current behavior

When I launch my E2E tests on Chrome, if I focus out of the Chrome window, the E2E commands are no longer running, it's hanging there doing nothing.

If I always stay focused on the Chrome window, the commands work as expected.

After focusing out, re-focusing in the window does not make the commands continue/restart.

Might be related to this PR this https://github.com/cypress-io/cypress/pull/28334

### Desired behavior

I should be able to run E2E tests while not being focused on Chrome (as before) so I can do something else while the E2E tests are running.

### Test code to reproduce

Sorry I don't have time to create a reproduction of the bug for the moment

### Cypress Version

13.6.0

### Node version

v18.18.2

### Operating System

macOS 14.1.1

### Debug Logs

_No response_

### Other

_No response_

### PR DESCRIPTION
**Pull request by cacieprins**
- Closes https://github.com/cypress-io/cypress/issues/28392

### Additional details

Previously, we were attempting to activate the main tab before and after every command. This required the Chrome extension to be loaded and enabled, and was only useful for users who were leveraging the Puppeteer plugin. If the extension was disabled or disallowed by policy, commands would cause the runner to hang even when the Puppeteer plugin was not in use.

This PR moves tab closing behavior into the Puppeteer plugin itself, and adds some support logic to the driver and extension in order to accommodate the logic happening in a Cypress plugin rather than in the browser.

### PR Tasks

- [X] Have tests been added/updated?
- [NA] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? 
- [NA] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?

### CODE REVIEW

File: cli/CHANGELOG.md [hunk 1]
```diff
@@ -7,6 +7,7 @@ _Released 2/13/2024 (PENDING)_
 
 - Fixed an issue which caused the browser to relaunch after closing the browser from the Launchpad. Fixes [#28852](https://github.com/cypress-io/cypress/issues/28852).
 - Fixed incorrect type definition of `dom.getContainsSelector`. Fixed in [#28339](https://github.com/cypress-io/cypress/pull/28339).
+- Fixed tests hanging when the Chrome browser extension is disabled. Fixes [#28392](https://github.com/cypress-io/cypress/issues/28392)
```
Comment (jennifer-shehane) [inline cli/CHANGELOG.md]: Try to order the changelog by most popular fixes at the top.

```suggestion
- Fixed tests hanging when the Chrome browser extension is disabled. Fixes [#28392](https://github.com/cypress-io/cypress/issues/28392)
- Fixed an issue which caused the browser to relaunch after closing the browser from the Launchpad. Fixes [#28852](https://github.com/cypress-io/cypress/issues/28852).
- Fixed incorrect type definition of `dom.getContainsSelector`. Fixed in [#28339](https://github.com/cypress-io/cypress/pull/28339).
```

File: npm/puppeteer/README.md [hunk 2]
```diff
@@ -333,6 +333,15 @@ export default defineConfig({
...
 ```
 
+## Troubleshooting
+
+### Error: Cannot communicate with the Cypress Chrome extension. Ensure the extension is enabled when using the Puppeteer plugin.
+
+If you receive this error in your command log, the Puppeteer plugin was unable to communicate with the Cypress extension. This extension is necessary in order to re-activate the main Cypress tab after a Puppeteer command, when running in open mode.
+
+* Ensure this extension is enabled in the instance of Chrome that Cypress launches by visiting chrome://extensions/
+* Ensure the Cypress extension is whitelisted by your company's security policy by its extension id, `caljajdfkjjjdehjdoimjkkakekklcck`
```
Comment (jennifer-shehane) [inline npm/puppeteer/README.md]: We avoid use of 'whitelist/blacklist' as a practice.

```suggestion
* Ensure the Cypress extension is allowed by your company's security policy by its extension id, `caljajdfkjjjdehjdoimjkkakekklcck`
```

File: npm/puppeteer/src/plugin/activateMainTab.ts [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+    onMessage = (ev) => {
+      if (ev.data.message === 'cypress:extension:main:tab:activated') {
+        window.removeEventListener('message', onMessage)
+        clearTimeout(timeout)
+        resolve()
+      }
+    }
+
+    window.addEventListener('message', onMessage)
+    window.postMessage({ message: 'cypress:extension:activate:main:tab' })
```
Comment (AtofStryker) [inline npm/puppeteer/src/plugin/activateMainTab.ts:26]: whats the difference between `cypress:extension:main:tab:activated` and `cypress:extension:activate:main:tab`?

**Review by AtofStryker — COMMENTED** [review]
looks like there is a dupe header in the changelog?

File: npm/puppeteer/src/plugin/activateMainTab.ts (hunk 3 above)
Comment (chrisbreiding) [reply npm/puppeteer/src/plugin/activateMainTab.ts:26]: `cypress:extension:activate:main:tab` is sent to the extension to tell it to activate the main tab. `cypress:extension:main:tab:activated` is the ack that the main tab was successfully activated.

**Review by chrisbreiding — APPROVED** [review]

**Review by AtofStryker — APPROVED** [review]

**Pull request merged**